File: PDOO/Napakalaki-Java/src/NapakalakiGame/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.*;
/**
 *
 * @author Usuario
 */
public class CardDealer {
    
    private static CardDealer instance = null;
    
    private ArrayList<Treasure> unusedTreasures=null;
    private ArrayList<Treasure> usedTreasures=null;
    private ArrayList<Monster> unusedMonsters=null;
    private ArrayList<Monster> usedMonsters=null;
    private ArrayList<Cultist> unusedCultists=null;
    private ArrayList<Cultist> usedCultists=null;
    
    
    public static CardDealer getInstance(){
        if(instance==null)
            instance=new CardDealer();
        return instance;
    }
    
    private CardDealer(){
        unusedTreasures=new ArrayList();
        usedTreasures=new ArrayList();
        unusedMonsters=new ArrayList();
        usedMonsters=new ArrayList();
        unusedCultists=new ArrayList();
        usedCultists=new ArrayList();
    }
    
    private void initCultistCardDeck(){
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",2));
        unusedCultists.add(new Cultist("Cultist",2));
        
        shuffleCultists();
    }
    
    private void initTreasureCardDeck(){
        Treasure T1 = new Treasure("¡Sí mi amo!", 4, TreasureKind.HELMET);
        unusedTreasures.add(T1);
        Treasure T2 = new Treasure("Botas de investigación", 3, TreasureKind.SHOE);
        unusedTreasures.add(T2);
        Treasure T3 = new Treasure("Capucha de Cthulhu", 3, TreasureKind.HELMET);
        unusedTreasures.add(T3);
        Treasure T4 = new Treasure("A prueba de babas", 2, TreasureKind.ARMOR);
        unusedTreasures.add(T4);
        Treasure T5 = new Treasure("Botas de lluvia ácida", 1, TreasureKind.SHOE);
        unusedTreasures.add(T5);
        Treasure T6 = new Treasure("Casco minero", 2, TreasureKind.HELMET);
        unusedTreasures.add(T6);
        Treasure T7 = new Treasure("Ametralladora Thompson", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T7);
        Treasure T8 = new Treasure("Camiseta de la UGR", 1, TreasureKind.ARMOR);
        unusedTreasures.add(T8);
        Treasure T9 = new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(T9);
        Treasure T10 = new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(T10);
        Treasure T11 = new Treasure("Fez alópodo", 3, TreasureKind.HELMET);
        unusedTreasures.add(T11);
        Treasure T12 = new Treasure("Hacha prehistórica", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(T12);
        Treasure T13 = new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR);
        unusedTreasures.add(T13);
        Treasure T14 = new Treasure("Gaita", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T14);
        Treasure T15 = new Treasure("Insecticida", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(T15);
        Treasure T16 = new Treasure("Escopeta de 3 cañones", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T16);
        Treasure T17 = new Treasure("Garabato místico", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(T17);
        Treasure T18 = new Treasure("La rebeca metálica", 2, TreasureKind.ARMOR);
        unusedTreasures.add(T18);
        Treasure T19 = new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T19);
        Treasure T20 = new Treasure("Necrocomicón", 1, TreasureKind.ONEHAND);
        unusedTreasures.add(T20);
        Treasure T21 = new Treasure("Necronomicón", 5, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T21);
        Treasure T22 = new Treasure("Linterna a 2 manos", 3, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T22);
        Treasure T23 = new Treasure("Necrognomicón", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(T23);
        Treasure T24 = new Treasure("Necrotelecom", 2, TreasureKind.HELMET);
        unusedTreasures.add(T24);
        Treasure T25 = new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(T25);
        Treasure T26 = new Treasure("Necroplayboycón", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(T26);
        Treasure T27 = new Treasure("Porra preternatural", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(T27);
        Treasure T28 = new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS);
        unusedTreasures.add(T28);
        Treasure T29 = new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(T29);
        Treasure T30 = new Treasure("Tentáculo de pega", 2, TreasureKind.HELMET);
        unusedTreasures.add(T30);
        Treasure T31 = new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOE);
        unusedTreasures.add(T31);  
        
        shuffleTreasures();
    }
    
    private void initMonsterCardDeck(){
        
        BadConsequence BC1 = new SpecificBadConsequence("Pierdes tu armadura visible y otra oculta", 0,
                             new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
                             new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
        Prize P1 = new Prize(2,1);
        unusedMonsters.add(new Monster("3 Byakhees de bonanza", 8, BC1, P1,0));
        
        BadConsequence BC2 = new SpecificBadConsequence("Embobados con el lindo primigenio te descartas "
                                 + "de tu casco visible", 0, 
                             new ArrayList(Arrays.asList(TreasureKind.HELMET)),
                             new ArrayList()); 
        Prize P2 = new Prize(1,1);
        unusedMonsters.add(new Monster("Chibithulhu", 2, BC2, P2,0));        
      
        BadConsequence BC3 = new SpecificBadConsequence("El primordial bostezo contagioso. Pierdes el"
                                 + " calzado visible", 0, 
                             new ArrayList(Arrays.asList(TreasureKind.SHOE)),
                             new ArrayList());
        Prize P3 = new Prize(1,1);
        unusedMonsters.add(new Monster("El sopor de Dunwich", 2, BC3, P3,0));
        
        BadConsequence BC4 = new SpecificBadConsequence("Te atrapan para llevarte de fiesta y te"
                             + " dejan caer\n en mitad del vuelo. Descarta 1 mano visible y"
                             + " 1 mano oculta", 0, 
                             new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),
                             new ArrayList(Arrays.asList(TreasureKind.ONEHAND)));
        Prize P4 = new Prize(4,1);
        unusedMonsters.add(new Monster("Angeles de la noche ibicenca", 14, BC4, P4,0));
        
        BadConsequence BC5 = new SpecificBadConsequence("Pierdes todos tus tesoros visibles", 0,
                             new ArrayList(Arrays.asList(TreasureKind.ARMOR, TreasureKind.BOTHHANDS,
                             TreasureKind.HELMET, TreasureKind.ONEHAND,
                             TreasureKind.SHOE)), new ArrayList());
        Prize P5 = new Prize(3,1);
        unusedMonsters.add(new Monster("El gorron en el umbral", 10, BC5, P5,0));
                        
        BadConsequence BC6 = new SpecificBadConsequence("Pierdes la armadura visible", 0, 
                             new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
                             new ArrayList());
        Prize P6 = new Prize(1,1);
        unusedMonsters.add(new Monster("H.P. Munchcraft", 6, BC6, P6,0));

        BadConsequence BC7 = new SpecificBadConsequence("Sientes bichos bajo la ropa. Descarta la"
                             + "armadura visible", 0, 
                             new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
                             new ArrayList());
        Prize P7 = new Prize(1,1);
        unusedMonsters.add(new Monster("Bichgooth", 2, BC7, P7,0));
        
        BadConsequence BC8 = new NumericBadConsequence("Pierdes 5 niveles y 3 tesoros visibles", 5, 3, 0);
        Prize P8 = new Prize(4,2);
        unusedMonsters.add(new Monster("El rey de rosa", 13, BC8, P8,0));
        
        BadConsequence BC9 = new NumericBadConsequence("Toses los pulmones y pierdes 2 niveles", 2, 0, 0);
        Prize P9 = new Prize(1,1);
        unusedMonsters.add(new Monster("La que redacta en las tinieblas", 2, BC9, P9,0));
        
        BadConsequence BC10 = new DeathBadConsequence("Estos monstruos resultan bastante superfic"
                             + "iales\n y te aburren mortalmente. Estas muerto");
        Prize P10 = new Prize(2,1);
        unusedMonsters.add(new Monster("Los hondos", 8, BC10, P10,0));
        
        BadConsequence BC11 = new NumericBadConsequence("Pierdes 2 niveles y 2 tesoros ocultos", 2, 0, 2);
        Prize P11 = new Prize(2,1);
        unusedMonsters.add(new Monster("Semillas Cthulhu", 4, BC11, P11,0));
        
        BadConsequence BC12 = new SpecificBadConsequence("Te intentas escaquear. Pierdes una mano visible.", 0, 
                             new ArrayList(Arrays.asList(TreasureKind.ONEHAND)), new ArrayList());
        Prize P12 = new Prize(2,1);
        unusedMonsters.add(new Monster("Dameargo", 1, BC12, P12,0));  
        
        BadConsequence BC13 = new NumericBadConsequence("Da mucho asquito. Pierdes 3 niveles.", 3, 0, 0);
        Prize P13 = new Prize(1,1);
        unusedMonsters.add(new Monster("Pollipolipo volante", 3, BC13, P13,0));
        
        BadConsequence BC14 = new DeathBadConsequence("No le hace gracia que pronuncien mal su nombre"
                               + ". Estás muerto.");
        Prize P14 = new Prize(3,1);
        unusedMonsters.add(new Monster("YskhtihyssgGoth", 12, BC14, P14,0));
         
        BadConsequence BC15 = new DeathBadConsequence("La familia te atrapa. Estás muerto.");
        Prize P15 = new Prize(4,1);
        unusedMonsters.add(new Monster("Familia feliz", 1, BC15, P15,0));
        
        BadConsequence BC16 = new SpecificBadConsequence("La quinta directiva primaria te obliga a \nperder 2 "
                              + "niveles y un tesoro 2 manos visible.", 2, 
                              new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)),
                              new ArrayList());
        Prize P16 = new Prize(2,1);
        unusedMonsters.add(new Monster("Roboggoth", 8, BC16, P16,0));
        
        BadConsequence BC17 = new SpecificBadConsequence("Te asusta en la noche. Pierdes un casco visible. ",0,
                              new ArrayList(Arrays.asList(TreasureKind.HELMET)),
                              new ArrayList());
        Prize P17 = new Prize(1,1);
        unusedMonsters.add(new Monster("El espia", 5, BC17, P17,0));
        
        BadConsequence BC18 = new NumericBadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 "
                              +" tesoros visibles. ", 2, 5, 0);
        Prize P18 = new Prize(1,1);
        unusedMonsters.add(new Monster("El lenguas", 20, BC18, P18,0));
        
        BadConsequence BC19 = new SpecificBadConsequence("Te faltan manos para tanta cabeza. \nPierdes 3 "
                              +"niveles y tus tesoros visibles de las manos. ", 3,
                              new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)),
                              new ArrayList());
        Prize P19 = new Prize(1,1);
        unusedMonsters.add(new Monster("Bicefalo", 20, BC19, P19,0));
        
        //CULTIST
        BadConsequence bc;
        Prize prize;
        
        ArrayList<TreasureKind> tVisible = new ArrayList();
        ArrayList<TreasureKind> tHidden = new ArrayList();
       
       tVisible.add(TreasureKind.ONEHAND);
       bc = new SpecificBadConsequence("Pierdes 1 mano visible",0,tVisible,null);
       prize = new Prize(3,1);
       unusedMonsters.add(new Monster("El mal indecible impronunciable",10,bc,prize,-2));
       
       tVisible = new ArrayList();
       
       bc = new NumericBadConsequence("Pierdes tus tesoros visibles. Jajaja.",0,BadConsequence.MAXTREASURES,0);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Testigos Oculares",6,bc,prize,+2));
       
       bc = new DeathBadConsequence("Hoy no es tu día de suerte. Mueres.");
       prize = new Prize(2,5);
       unusedMonsters.add(new Monster("El gran chulhu",20,bc,prize,+4));
       
       bc = new NumericBadConsequence("Tu gobierno te recorta 2 niveles.",2,0,0);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Serpiente Político",8,bc,prize,-2));
       
       tVisible.add(TreasureKind.ARMOR);
       tVisible.add(TreasureKind.HELMET);
       tHidden.add(TreasureKind.BOTHHANDS);
       bc = new SpecificBadConsequence("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.",0,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Felpuggoth",2,bc,prize,+5));
       
       bc = new NumericBadConsequence("Pierdes 2 niveles.",2,0,0);
       prize = new Prize(4,2);
       unusedMonsters.add(new Monster("Shoggoth",16,bc,prize,-4));
       
       bc = new NumericBadConsequence("Pintalabios negro. Pierdes 2 niveles.",2,0,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Lolitagooth",2,bc,prize,+3));
       
       shuffleMonsters();
    }
    
    private void shuffleTreasures(){
       Collections.shuffle(unusedTreasures);
    }
    
    private void shuffleMonsters(){
        Collections.shuffle(unusedMonsters);
    }
    
    private void shuffleCultists(){
        Collections.shuffle(unusedCultists);
    }
    
    public Treasure nextTreasure(){
        if(unusedTreasures.isEmpty()){
            unusedTreasures=usedTreasures;
            usedTreasures=new ArrayList();
            shuffleTreasures();
        }
        Treasure nextTreasure = unusedTreasures.get(0);
        unusedTreasures.remove(nextTreasure);
        usedTreasures.add(nextTreasure);
        return nextTreasure;        
    }
    
    public Monster nextMonster(){
        if(unusedMonsters.isEmpty()){
            unusedMonsters=usedMonsters;
            usedMonsters=new ArrayList();
            shuffleMonsters();
        }
        Monster nextMonster = unusedMonsters.get(0);
        unusedMonsters.remove(nextMonster);
        usedMonsters.add(nextMonster);
        return nextMonster;   
    }
    
    public Cultist nextCultist(){
        if(unusedCultists.isEmpty()){
            unusedCultists=usedCultists;
            usedCultists=new ArrayList();
            shuffleCultists();
        }
        Cultist nextCultist = unusedCultists.get(0);
        unusedCultists.remove(nextCultist);
        usedCultists.add(nextCultist);
        return nextCultist;      
    }
    
    public void giveTreasureBack(Treasure t){
        usedTreasures.add(t); 
        unusedTreasures.remove(t);
    }
    
    public void giveMonsterBack(Monster m){
        usedMonsters.add(m);
        unusedMonsters.remove(m);
    }
    
    public void initCards(){
        this.initTreasureCardDeck();
        this.initMonsterCardDeck();
        this.initCultistCardDeck();
    }
}

File: PDOO/Napakalaki-Ruby/pkg/Napakalaki-Ruby-0.0.1/lib/CardDealer.rb
#encoding: utf-8

require 'singleton'

require 'Treasure.rb'
require 'TreasureKind.rb'
require 'BadConsequence.rb'
require 'NumericBadConsequence.rb'
require 'SpecificBadConsequence.rb'
require 'DeathBadConsequence.rb'
require 'Napakalaki.rb'
require 'Player.rb'
require 'Dice.rb'
require 'Monster.rb'
require 'Prize.rb'
require 'Cultist.rb'


module NapakalakiGame

class CardDealer
    
    include Singleton 
    
    attr_reader :unusedMonsters, :usedMonsters, :unusedTreasures, :usedTreasures, :unusedCultists, :usedCultists
    
    def initialize
        @unusedMonsters=Array.new
        @unusedTreasures=Array.new
        @usedMonsters=Array.new
        @usedTreasures=Array.new
        @unusedCultists=Array.new
        @usedCultists=Array.new
    end
    
    private
    
    def initCultistCardDeck()
        @unusedCultists << Cultist.new("Cultist",1)
        @unusedCultists << Cultist.new("Cultist",1)
        @unusedCultists << Cultist.new("Cultist",1)
        @unusedCultists << Cultist.new("Cultist",1)
        @unusedCultists << Cultist.new("Cultist",2)
        @unusedCultists << Cultist.new("Cultist",2)
    end
    
    def initTreasureCardDeck
        tesoro = Treasure.new("¡Sí mi amo!", 4, TreasureKind::HELMET)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Botas de investigación", 3, TreasureKind::SHOE)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Capucha de Cthulhu", 3, TreasureKind::HELMET)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("A prueba de babas", 2, TreasureKind::ARMOR)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Botas de lluvia ácida", 1, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Casco minero", 2, TreasureKind::HELMET)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Ametralladora Thompson", 4, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Camiseta de la UGR", 1, TreasureKind::ARMOR)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Clavo de rail ferroviario", 3, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Cuchillo de sushi arcano", 2, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Fez alópodo", 3, TreasureKind::HELMET)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Hacha prehistórica", 2, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("El aparato del Pr. Tesla", 4, TreasureKind::ARMOR)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Gaita", 4, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Insecticida", 2, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Escopeta de 3 cañones", 4, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Garabato místico", 2, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("La rebeca metálica", 2, TreasureKind::ARMOR)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Lanzallamas", 4, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Necrocomicón", 1, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Necronomicón", 5, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Linterna a 2 manos", 3, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Necrognomicón", 2, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Necrotelecom", 2, TreasureKind::HELMET)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Mazo de los antiguos", 3, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Necroplayboycón", 3, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Porra preternatural", 2, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Shogulador", 1, TreasureKind::BOTHHANDS)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Varita de atizamiento", 3, TreasureKind::ONEHAND)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Tentáculo de pega", 2, TreasureKind::HELMET)
        @unusedTreasures << tesoro

        tesoro = Treasure.new("Zapato deja-amigos", 1, TreasureKind::SHOE)
        @unusedTreasures << tesoro 
    end
    
    def initMonsterCardDeck
        pr = Prize.new(2,1)
        bc = NumericBadConsequence.new("Pierdes tu armadura visible y otra oculta", 0, 1, 1)
        @unusedMonsters << Monster.Monster("3 Byakhees de bonanza", 8, bc, pr)

        pr = Prize.new(1, 1)
        bc = SpecificBadConsequence.new("Embobados con el lindo primigenio te descartas de tu casco visible", 0, 
                TreasureKind::HELMET, Array.new)
        @unusedMonsters << Monster.Monster("Chibithulhu", 2, bc, pr)     

        pr = Prize.new(1, 1)
        bc = SpecificBadConsequence.new("El primordial bostezo contagioso. Pierdes el calzado visible. ", 0,
                    TreasureKind::SHOE, Array.new)
        @unusedMonsters<< Monster.Monster("El sopor de Dunwich", 2, bc, pr)

        pr = Prize.new(4, 1)
        bc = SpecificBadConsequence.new("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. 
                  Descarta 1 mano visible y 1 mano oculta", 0,  TreasureKind::ONEHAND, TreasureKind::ONEHAND)
        @unusedMonsters << Monster.Monster("Ángeles de la noche ibicenca ", 14, bc ,pr)

        pr = Prize.new(3, 1)
        bc = NumericBadConsequence.new("Pierdes todos tus tesoros visibles. ", 0, 100, 0)
        @unusedMonsters << Monster.Monster("El gorrón en el umbral ", 10, bc ,pr)

        pr = Prize.new(2, 1)
        bc = SpecificBadConsequence.new("Pierdes la armadura visible. ",0 , TreasureKind::ARMOR,  Array.new)
        @unusedMonsters << Monster.Monster("H.P Munchcraft ", 6, bc ,pr)

        pr = Prize.new(1, 1)
        bc = SpecificBadConsequence.new("Sientes bichos bajo la ropa. Descarta la armadura visible. ", 0, 
                  TreasureKind::ARMOR,  Array.new)
        @unusedMonsters << Monster.Monster("Bichgooth " , 2, bc ,pr)

        pr = Prize.new(4, 2)
        bc = NumericBadConsequence.new("Pierdes 5 niveles y 3 tesoros visibles. ", 5, 3, 0)
        @unusedMonsters << Monster.Monster("El rey de rosa ", 13, bc ,pr)

        pr = Prize.new(1, 1)
        bc = NumericBadConsequence.new("Toses los pulmones y pierdes 2 niveles. ", 2, 0, 0)
        @unusedMonsters << Monster.Monster("La que redacta en las tinieblas ", 2, bc ,pr)

        pr = Prize.new(2, 1)
        bc = DeathBadConsequence.new("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estás muerto. ")
        @unusedMonsters << Monster.Monster("Los hondos ", 8, bc, pr)

        pr = Prize.new(2, 1)
        bc = NumericBadConsequence.new("Pierdes 2 niveles y 2 tesoros ocultos. ", 2, 0, 2)
        @unusedMonsters << Monster.Monster("Semillas Cthulhu ", 4, bc ,pr)

        pr = Prize.new(2, 1)
        bc = SpecificBadConsequence.new("Te intentas escaquear. Pierdes una mano visible. " , 0, 
                      TreasureKind::ONEHAND,  Array.new)
        @unusedMonsters << Monster.Monster("Dameargo ", 1, bc ,pr)

        pr = Prize.new(1, 1)
        bc = BadConsequence.new("Da mucho asquito. Pierdes 3 niveles. ", 3)
        @unusedMonsters << Monster.Monster("Pollipólipo volante ", 3, bc ,pr)

        pr = Prize.new(3, 1)
        bc = DeathBadConsequence.new("No le hace gracia que pronuncien mal su nombre. Estás muerto. ")
        @unusedMonsters << Monster.Monster("Yskhtihyssg-Goth ", 12, bc ,pr)

        pr = Prize.new(4, 1)
        bc = DeathBadConsequence.new("La familia te atrapa. Estás muerto. ")
        @unusedMonsters << Monster.Monster("Familia feliz ", 1, bc ,pr)

        pr = Prize.new(2, 1)
        bc = SpecificBadConsequence.new("La quinta directiva primaria te obliga a perder 2 niveles y un
                    tesoro 2 manos visible. ", 2, TreasureKind::BOTHHANDS,  Array.new)
        @unusedMonsters << Monster.Monster("Roboggoth ", 8, bc ,pr)

        pr = Prize.new(1, 1)
        bc  = SpecificBadConsequence.new("Te asusta en la noche. Pierdes un casco visible. ", 0, 
                    TreasureKind::HELMET,  Array.new)
        @unusedMonsters << Monster.Monster("El espía ", 5, bc,pr)

        pr = Prize.new(1, 1)
        bc = NumericBadConsequence.new("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles. ", 2, 5, 0)
        @unusedMonsters << Monster.Monster("El lenguas ", 20, bc ,pr)

        pr = Prize.new(1, 1)
        bc = SpecificBadConsequence.new("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos. ", 
                    3, TreasureKind::BOTHHANDS,  Array.new)
        @unusedMonsters << Monster.Monster("Bicéfalo ", 20, bc ,pr)

          #CULTISTS

        pr = Prize.new(2,1)
        bc = NumericBadConsequence.new("Pierdes tus tesoros visibles. Jajaja.",0,BadConsequence.MAXTREASURES, 0)
        @unusedMonsters << Monster.MonsterC("Testigos Oculares",6,bc,pr,+2)

        pr = Prize.new(2,5)
        bc = DeathBadConsequence.new("Hoy no es tu día de suerte. Mueres.")
        @unusedMonsters << Monster.MonsterC("El gran cthulhu",20,bc,pr,+4)

        pr = Prize.new(2,1)
        bc = NumericBadConsequence.new("Tu gobierno te recorta 2 niveles.", 2, 0,0)
        @unusedMonsters << Monster.MonsterC("Serpiente Político",8,bc,pr,-2)

        pr = Prize.new(1,1)
        bc = SpecificBadConsequence.new("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.",0,[TreasureKind::ARMOR,TreasureKind::HELMET],[TreasureKind::BOTHHANDS])
        @unusedMonsters << Monster.MonsterC("Felpuggoth",2,bc,pr,+5)

        pr = Prize.new(4,2)
        bc = NumericBadConsequence.new("Pierdes 2 niveles",2,0,0)
        @unusedMonsters << Monster.MonsterC("Shoggoth",16,bc,pr,-4)

        pr = Prize.new(1,1)
        bc = NumericBadConsequence.new("Pintalabios negro. Pierdes 2 niveles.",2,0,0)
        @unusedMonsters << Monster.MonsterC("Lolitagooth",2,bc,pr,+3)
    end
    
    def shuffleTreasures
        @unusedTreasures.shuffle!
    end
    
    def shuffleMonsters
        @unusedMonsters.shuffle!
    end
    
    def shuffleCultists
      @unusedCultists.shuffle!
    end
    
    
    #------------------------ PARTE PÚBLICA  ------------------------
    
    public 
    
    
    def nextTreasure
        if(@unusedTreasures.length() == 0)
        @usedTreasures.each do |t|
          @unusedTreasures << @usedTreasures.delete(t)
        end
        shuffleTreasures
      end
      return @unusedTreasures.delete_at(0)
    end
    
    def nextMonster
        if(@unusedMonsters.length() == 0)
        @usedMonster.each do |m|
          @unusedMonsters << @usedMonster.delete(m)
        end
        shuffleMonsters
      end
      return @unusedMonsters.delete_at(0)
    end
    
    def nextCultist
      return @unusedCultists.delete_at(0)
    end
    
    def giveTreasureBack(t)
        @usedTreasures << t
    end
    
    def giveMonsterBack(m)
        @usedMonsters << m
    end
    
    def initCards
        initTreasureCardDeck
        initMonsterCardDeck
        initCultistCardDeck
        shuffleTreasures
        shuffleMonsters
        shuffleCultists
    end
    
 end   
end
